Automattic/jetpack
# Jetpack: check for other plugins before uninstall cleanup.

## Proposed changes:
* Introduce new `Manager::is_ready_for_cleanup()` method in Connection package to check if a plugin should perform the cleanup, or there are other plugins using connection.
* Make use of that new method on Jetpack uninstall to keep connection intact for other plugins when needed.

### Other information:


## Jetpack product discussion
p1698683954801629-slack-CBG1CP4EN

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
Testing is tricky because we can't use Jetpack Beta for uninstall testing.
I suggest using JN.
You can use release versions of Jetpack and Jetpack Backup, but you'll need to run these commands to overwrite the changed files.

Jetpack (from the site's root directory):
```bash
curl https://raw.githubusercontent.com/Automattic/jetpack/update/jetpack-no-disconnect-on-uninstall/projects/packages/connection/src/class-manager.php > wp-content/plugins/jetpack/jetpack_vendor/automattic/jetpack-connection/src/class-manager.php
curl https://raw.githubusercontent.com/Automattic/jetpack/update/jetpack-no-disconnect-on-uninstall/projects/plugins/jetpack/uninstall.php > wp-content/plugins/jetpack/uninstall.php
```

Jetpack Backup:
```bash
curl https://raw.githubusercontent.com/Automattic/jetpack/update/jetpack-no-disconnect-on-uninstall/projects/packages/connection/src/class-manager.php > wp-content/plugins/jetpack-backup/jetpack_vendor/automattic/jetpack-connection/src/class-manager.php
```

1. Install and connect Jetpack, overwrite the files (see above).
2. Install a standalone plugin (e.g. Jetpack VaultPress Backup), purchase a plan. Overwrite the `class-manager.php` (see above).
3. Go to "Plugins", deactivate and uninstall Jetpack.
4. Confirm VaultPress Backup is still connected.
5. Deactivate VaultPress Backup, reinstall and reconnect Jetpack. Overwrite the files (see above).
6. Go to "Plugins" and deactivate and uninstall Jetpack.
7. Confirm the options were cleaned up (you can use `wp option get jetpack_options`).
